### PR TITLE
cabana: sort events based on priority

### PR DESCRIPTION
**Pull request by jnewb1**
**Description**

Sorts events on the cabana timeline based on some order that I thought made sense. This allows things like user flags to actually appear on the timeline (because they will be painted last).

image

### CODE REVIEW

File: tools/cabana/videowidget.cc [hunk 1]
```diff
@@ -183,13 +183,33 @@ void Slider::sliderChange(QAbstractSlider::SliderChange change) {
   }
 }
 
+bool sortTimelineBasedOnEventPriority(std::tuple<int, int, TimelineType> left, std::tuple<int, int, TimelineType> right){
+  std::map<TimelineType, int> timelinePriority = {
```
Comment (deanlee) [inline tools/cabana/videowidget.cc:187]: timelinePriority can be` const static`

File: tools/cabana/videowidget.cc [hunk 2]
```diff
@@ -183,13 +183,33 @@ void Slider::sliderChange(QAbstractSlider::SliderChange change) {
...
+
 void Slider::paintEvent(QPaintEvent *ev) {
   QPainter p(this);
   QRect r = rect().adjusted(0, 4, 0, -4);
   p.fillRect(r, timeline_colors[(int)TimelineType::None]);
   double min = minimum() / 1000.0;
   double max = maximum() / 1000.0;
-  for (auto [begin, end, type] : timeline) {
+
+  std::vector<std::tuple<int, int, TimelineType>> sortedTimeline;
```
Comment (deanlee) [inline tools/cabana/videowidget.cc:206]: We don't need to sort the same timeline repeatedly in paintEvent. think it's better to sort it in 
```
timer.callOnTimeout([this]() {
    timeline = can->getTimeline();
    update();
  });
```

**Comment by deanlee** [comment]
very nice.thank you!

File: tools/cabana/videowidget.cc (hunk 2 above)
Comment (jnewb1) [reply tools/cabana/videowidget.cc:206]: done

File: tools/cabana/videowidget.cc (hunk 1 above)
Comment (jnewb1) [reply tools/cabana/videowidget.cc]: done

File: tools/cabana/videowidget.cc [hunk 3]
```diff
@@ -128,7 +142,16 @@ void VideoWidget::updatePlayBtnState() {
...
-    timeline = can->getTimeline();
+    auto unsortedTimeline = can->getTimeline();
+
+    std::vector<std::tuple<int, int, TimelineType>> sortedTimeline;
+
+    std::copy(unsortedTimeline.begin(), unsortedTimeline.end(), std::back_inserter(sortedTimeline));
+
+    std::sort(sortedTimeline.begin(), sortedTimeline.end(), sortTimelineBasedOnEventPriority);
+
+    timeline = sortedTimeline;
```
Comment (deanlee) [inline tools/cabana/videowidget.cc:153]: can simplify to:
timeline = can->getTimeline();
std::sort(timeline.begin(), timeline.end(), sortTimelineBasedOnEventPriority);

File: tools/cabana/videowidget.cc [hunk 4]
```diff
@@ -19,6 +19,20 @@ static const QColor timeline_colors[] = {
   [(int)TimelineType::AlertCritical] = QColor(199, 0, 57),
 };
 
+
+bool sortTimelineBasedOnEventPriority(std::tuple<int, int, TimelineType> left, std::tuple<int, int, TimelineType> right){
```
Comment (deanlee) [inline tools/cabana/videowidget.cc:23]: pass left&right by const referer

File: tools/cabana/videowidget.cc [hunk 5]
```diff
@@ -128,7 +142,9 @@ void VideoWidget::updatePlayBtnState() {
 // Slider
 Slider::Slider(QWidget *parent) : timer(this), thumbnail_label(this), QSlider(Qt::Horizontal, parent) {
   timer.callOnTimeout([this]() {
-    timeline = can->getTimeline();
+    auto timeline = can->getTimeline();
```
Comment (deanlee) [inline tools/cabana/videowidget.cc:145]: timeline = can->getTimeline();

**Comment by deanlee** [comment]
lgtm

**Comment by adeebshihadeh** [comment]
Thanks for the PR!

**Pull request merged**